x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportPutShutdownNodeAction.java

```diff
@@ -126,6 +129,7 @@ public ClusterState execute(BatchExecutionContext<PutShutdownNodeTask> batchExec
                 return updatedState;
             }
 
+            resetDesiredBalance.run();
```

Comment:
makes sense. thanks for the feedback.

Reply:
I guess, rather than doing that check whenever the reroute happens (`allocate` gets called) , the allocator should listen for clusterState changes and just do the reset (followed by a new reroute)?

Reply:
Yeah we need to track some extra state in the allocator against which to compare. That's fine tho.

Reply:
ok. It is not clear to me why reacting on the CS change is not preferable. Are we going to reuse that extra state for something else in the allocator?

Reply:
No I wouldn't expect so.

The allocator sorta runs separately from cluster state updates - it consumes cluster states, but a little asynchronously (and possibly more asynchronously in future). IMO `org.elasticsearch.cluster.ClusterChangedEvent#previousState` is a bit of an antipattern, it forces every component to synchronously consume every state update.

Also there's no need to even look at most state updates, only the ones that involve a reroute, and I could imagine that we end up with some logic in this area that we'd rather avoid running on every state update. Better to keep hold of a few objects out of the last cluster state we processed instead.

That said, I'm not dead against watching for cluster state updates here, it's not how I'd do it but if you think it's simpler then go for it.

Reply:
Thanks for explaining. Just wanted to understand the pros/cons. Sounds very reasonable to me! :)